common/quotas/global/rpc/client.go

```diff
@@ -0,0 +1,191 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// Package rpc contains a concurrent RPC client, and handles mapping to/from the
+// Any-typed request details so other packages do not have to concern themselves
+// with those details.
+package rpc
+
+//go:generate mockgen -package $GOPACKAGE -source $GOFILE -destination client_mock.go -self_package github.com/uber/cadence/common/quotas/global/rpc Client
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"go.uber.org/yarpc"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/uber/cadence/client/history"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/types"
+)
+
+type (
+	Calls struct {
+		Allowed  int
+		Rejected int
+	}
+)
+
+type (
+	// UpdateResult holds all successes and errors encountered, to be distinct
+	// from fatal errors that imply unusable results (i.e. a normal error return).
+	UpdateResult struct {
+		// all successfully-returned weights.
+		// this may be populated and is safe to use even if Errors is present,
+		// as some shards may have succeeded.
+		Weights map[string]float64
+		// Any unexpected errors encountered, or nil if none.
+		// Weights is valid even if this is present, though it may be empty.
+		//
+		// This should always be nil aside from major programming or rollout
+		// errors, e.g. failures to serialize or deserialize data (coding errors
+		// or incompatible server versions).
+		Err error
+	}
+
+	Client interface {
+		// Update performs concurrent calls to all aggregating peers to send load info
+		// and retrieve new per-key weights from the rest of the cluster.
+		// This is intended to be called periodically in the background per process,
+		// not synchronously or concurrently (per set of keys anyway), as it is fairly
+		// high cost to shard keys and send so many requests.
+		//
+		// Currently, there are no truly fatal errors so this API does not return `error`.
+		// Even in the presence of errors, some successful data may have been loaded,
+		// and that will be part of the UpdateResult struct.
+		Update(ctx context.Context, period time.Duration, load map[string]Calls) UpdateResult
+	}
+
+	client struct {
+		history  history.Client
+		resolver history.PeerResolver
+		thisHost string
+
+		logger log.Logger
+		scope  metrics.Scope
+	}
+)
+
+var _ Client = (*client)(nil)
+
+func New(
+	historyClient history.Client,
+	resolver history.PeerResolver,
+	logger log.Logger,
+	met metrics.Client,
+) Client {
+	return &client{
+		history:  historyClient,
+		resolver: resolver,
+		thisHost: uuid.NewString(), // TODO: would descriptive be better?  but it works, unique ensures correctness.
```

Comment:
Not entirely sure what the value of a random string is there over a host/container instance name? I find this a little odd, do we not have a way to get runtime identity for our other clients?

Reply:
> do we not have a way to get runtime identity for our other clients?

I'm honestly not sure.  I couldn't find anything conclusive or even describing an attempt at a per-process unique ID.

I suspect there _is_ one in the membership/peer/etc ringpop stuff, but I don't know the intricacies there well enough to figure out what it would be.

re container restarts: it is _potentially_ fine since that will _potentially_ result in the new process receiving the same pattern of requests it did before...... but there really isn't any way to know if that's true in general.  

though the consequences either way are quite minor, and look pretty similar whether the host-aggregated-data is reused or lost.  the only bit that's actually important is that two _active_ hosts never choose the same value, because they won't weight fairly.

---

that said: this can be changed at any time, since it'll also change on every deploy.  so if we do find a nicer value, it'll be trivial to adopt.

Reply:
To avoid confusion you can use the hostname that's available in service resource object.  Similar to uuid that hostname will change with restarts/deployments which is fine. it's just a unique identifier of the service instance. 
Existing implementation lacks debug logs but when/if we introduce them it would be nice to avoid a random uuid and see hostname there.

Reply:
I don't actually see that value being set anywhere at all 🤔 it's read, but never assigned.

If it's _literally_ the host name, like `os.Hostname()`, that's not good enough because it would mean multiple instances on the same machine share the same name.

Reply:
to stick IRL chat in here:
gonna stick with UUID for now.  but I agree something from the ring would probably be preferable, if we can find something that is truly unique and stable.

there _should_ be something in there, our rings depend on it, but I'm not confident that it's `host:port` (though that is currently unique internally) instead of some other field we might not currently be exposing.  if/when we find it, switching makes plenty of sense as it should be a more shared / identifiable value.